App History to Navigation API (#2897)
GoogleChrome/developer.chrome.com: site/en/docs/web-platform/navigation-api/index.md

---
layout: 'layouts/doc-post.njk'
title: 'Modern client-side routing: the Navigation API'
subhead: 'Standardizing client-side routing through a brand new API which completely overhauls building single-page applications.'
authors:
  - samthor
date: 2021-08-25

> updated: 2021-09-20

Reviewer: ```suggestion
updated: 2022-05-26
```

This event is fundamentally _centralized_: it will fire for all types of navigations, whether the user performed an action (such as clicking a link, submitting a form, or going back and forward) or when navigation is triggered programmatically (i.e., via your site's code).
In most cases, it lets your code override the browser's default behavior for that action.
For SPAs, that likely means keeping the user on the same page and loading or changing the site's content.

A `NavigateEvent` is passed to the "navigate" listener which contains information about the navigation, such as the destination URL, and allows you to respond to the navigation in one centralized place.
A basic "navigate" listener on example.com could look like this:

```js
navigation.addEventListener('navigate', navigateEvent => {
  switch (navigateEvent.destination.url) {
    case 'https://example.com/':
      navigateEvent.transitionWhile(loadIndexPage());
      break;
    case 'https://example.com/cats':
      navigateEvent.transitionWhile(loadCatsPage());
      break;
  }
});
```

You can intercept the navigation in one of two ways:

> - by calling `transitionWhile()` (as described above) to handle the navigation
> - by calling `preventDefault()`, which can cancel the navigation completely

Reviewer: ```suggestion
- Calling `transitionWhile()` (as described above) to handle the navigation.
- Calling `preventDefault()`, which can cancel the navigation completely.
```

This example calls `transitionWhile()` on the event with a promise generated from async functions.
By calling this method, the browser knows that your code will configure the next state of your site.
This will create a transition object, `navigation.transition`, which other code can use to track the progress of the transition.

Both `transitionWhile()` and `preventDefault()` are usually allowed, but have cases where they're unable to be called.

> You can't handle navigations via `transitionWhile()` if the navigation is a cross-origin navigation: e.g., if it's leaving your domain.

Reviewer: ```suggestion
You can't handle navigations via `transitionWhile()` if the navigation is a cross-origin navigation, for example, if it's leaving your domain.
```

Reviewer: style guide suggests avoiding e.g. and i.e.

And you can't cancel a navigation via `preventDefault()` if the user is pressing the Back or Forward buttons in their browser; you should not be able to trap your users on your site.
(This is [being discussed on GitHub][back-forward-discuss].)

> Even if you can't stop or intercept the navigation itself, the "navigate" event will still fire.
> It's _informative_, so your code could, e.g., log an Analytics event to indicate that a user is leaving your site.

Reviewer: ```suggestion
Even if you can't stop or intercept the navigation itself, the "navigate" event will still fire.
It's _informative_, so your code could, for example, log an Analytics event to indicate that a user is leaving your site.
```

These issues are being [discussed on GitHub](https://github.com/WICG/navigation-api/issues/66).
{% endAside %}

When you pass a promise to `transitionWhile()`, one of two things will happen:

- If that `Promise` fulfills (or you did not call `transitionWhile()`), the Navigation API will fire "navigatesuccess" with an `Event`.
- If that `Promise` rejects, the API will fire "navigateerror" with an `ErrorEvent`.

These events allow your code to deal with success or failure in a centralized way.
For example, you might deal with success by hiding a previously displayed progress indicator, like this:

```js
navigation.addEventListener('navigatesuccess', event => {
  loadingIndicator.hidden = true;
});
```

Or you might show an error message on failure (i.e., if the `Promise` passed to `transitionWhile` rejected):

```js
navigation.addEventListener('navigateerror', event => {
  loadingIndicator.hidden = true; // also hide indicator
  showMessage(`Failed to load page: ${event.message}`);
});
```

The "navigateerror" event listener, which receives an `ErrorEvent`, is particularly handy as it's guaranteed to receive any errors from your code that's setting up a new page.

> You can simply `await fetch()` knowing that if the network is unavailable, the error will eventually be routed to "navigateerror".

Reviewer: ```suggestion
You can simply `await fetch()` knowing that if the network is unavailable, the error will eventually be routed to `"navigateerror"`.
```

Your page might deal with these, but there's a long tail of possibilities which could just be simplified—something that the new Navigation API achieves.

Personally, the History API often _feels_ like it could go some way to help with these possibilities.
However, it really only has two surface areas: responding if the user presses Back or Forward in their browser, plus pushing and replacing URLs.
It doesn't have an analogy to "navigate", except if you manually set up listeners for, e.g., click events, as demonstrated above.

## Transition

When your code calls `transitionWhile()` from within its "navigate" listener, it informs the browser that it's now preparing the page for the new, updated state; and that the navigation may take some time. The `Promise` you pass to `transitionWhile()` tells the browser how long the navigation takes.

As such, this API introduces a semantic concept that the browser understands: an SPA navigation is currently occurring, over time, changing the document from a previous URL and state to a new one.
This has a number of potential benefits, including accessibility: browsers can surface the beginning, end, or potential failure of a navigation.

> Chrome, for example, activates its native loading indicator, and allows the user to interact with the stop button. (This doesn't currently happen when the user navigates via the back/forward buttons, but that (will be fixed soon)[loading-crbug].)

Reviewer 2: This link syntax looks suspicious?

Author: Good catch (I don't like these footnote style links, so I'm not used to using them)

### Transition Success and Failure

After the "navigate" event completes, the URL being navigated to will take effect.
This happens immediately, even if you've called `transitionWhile()`.

{% Aside 'caution' %}

> This means if there's a delay updating the page, there's a period where the old content is being displayed with the new URL. This can change the behavior of relative URLs (not just links, also other URLs like images and scripts), and other scripts that rely on the current history entry to represent current content (such as `location.href`). These issues are [being discussed on GitHub](https://github.com/WICG/navigation-api/issues/232).

Reviewer 2: "not just links, also other URLs like images and scripts": well, it doesn't change scripts and images that are already loaded... I think this is a bit misleading. Maybe "not just links, also any new subresources you load"?

{% endAside %}

> When you pass a promise to `transitionWhile`, one of two things will happen:

Reviewer 2: ```suggestion
When you pass a promise to `transitionWhile()`, one of two things will happen:
```

?

These events allow your code to deal with success or failure in a centralized way.
For example, you might deal with success by hiding a previously displayed progress indicator, like this:

```js
navigation.addEventListener('navigatesuccess', event => {
  loadingIndicator.hidden = true;
});
```

Or you might show an error message on failure (i.e., if the `Promise` passed to `transitionWhile` rejected):

```js
navigation.addEventListener('navigateerror', event => {
  loadingIndicator.hidden = true; // also hide indicator
  showMessage(`Failed to load page: ${event.message}`);
});
```

The "navigateerror" event listener, which receives an `ErrorEvent`, is particularly handy as it's guaranteed to receive any errors from your code that's setting up a new page.
You can simply `await fetch()` knowing that if the network is unavailable, the error will eventually be routed to "navigateerror".

### Abort Signals

Since you're able to do asynchronous work while preparing a new page, it's possible that the transition your code is handling (to load a specific URL or state) might get preempted, or considered out-of-date.

> This might happen because the user just clicked on another link, or your code itself throws an error.

Reviewer 2: I'm actually not sure "your code itself throws an error" is something that can cause this... maybe "your code itself performs another navigation"?

Author: huh, yeah. Was that link always wrong, or was it a change to the spec?

Reviewer 2: I think that was always wrong, although maybe it was referring to how a transition is considered a failure if the IIAFE throws an error. (Failure is different than preemption, but this is the closest I can think of.)

To deal with any of these possibilities, the event passed to the "navigate" listener contains a `signal` property, which is an `AbortSignal`.
For more information see [Abortable fetch][abortable-fetch].
The short version is it basically provides an object that fires an event when you should stop your work.
Notably, you can pass an `AbortSignal` to any calls you make to `fetch()`, which will cancel in-flight network requests if the navigation is preempted.
This will both save the user's bandwidth, and reject the `Promise` returned by `fetch()`, preventing any following code from e.g., updating the DOM to show a now invalid page navigation.

For a concrete example, you might set up loading a page of cat memes with a `fetch()` call in your listener.
By passing the `signal` to it, the fetch will be cancelled if the user decides to instead load a different page on your site before the `fetch` completes.
Take a look:

```js
navigation.addEventListener('navigate', navigateEvent => {
  if (isCatsUrl(navigateEvent.destination.url)) {

>     const method = async () => {

Reviewer 2: Preexisting issue... this is more of a function than a method... maybe name it something like processNavigation.

      const request = await fetch('/cat-memes.json', {
        signal: navigateEvent.signal,
      });
      const json = await request.json();
      // TODO: do something with cat memes json
    };
    navigateEvent.transitionWhile(method());
  } else {
    // load some other page
  }
});
```

## Navigation Entries

> `navigation.current` provides access to the current entry.

Reviewer 2: current -> currentEntry. Maybe do a Ctrl+F to make sure you caught them all...

Author: Good catch

```

By default, this will be `undefined`.
You can synchronously set the state for the current `NavigationEntry` by calling:

```js
navigation.updateCurrent({state: something});
```

You can also set the state when navigating programmatically with `navigation.navigate()` (this is [described below](#programmatic-navigation)).

In the Navigation API, the state returned from `.getState()` is a copy of the previously set state.
If you modify it, the stored version won't also change.
For example:

```js
navigation.updateCurrent({state: {count: 1}});

const state = navigation.current.getState();
state.count = 2;

console.info(navigation.current.getState().count); // will still be one
```

### Access All Entries

The "current entry" is not all, though.
The API also provides a way to access the entire list of entries that a user has navigated through while using your site via its `navigation.entries()` call, which returns a snapshot array of entries.
This could be used to, e.g., show a different UI based on how the user navigated to a certain page, or just to look back at the previous URLs or their states.
This is impossible with the current History API.

> You can also listen for a "dispose" event on individual `NavigationEntry`s, which is fired when the entry is no longer part of browser history. This can happen as part of general cleanup, but also happen when navigating. For example, if you traverse back 10 places, then navigate forwards, those 10 history entries will be disposed.

Reviewer 2: ```suggestion
You can also listen for a "dispose" event on individual `NavigationHistoryEntry`s, which is fired when the entry is no longer part of browser history. This can happen as part of general cleanup, but also happen when navigating. For example, if you traverse back 10 places, then navigate forwards, those 10 history entries will be disposed.
```

here and throughout

Author: Didn't realise this had changed. Will change the other references too

This is intended as an improved aggregation of older methods like `location.assign()` and friends, plus the History API's methods `pushState()` and `replaceState()`.

{% Aside %}

These older programmatic methods for changing the URL are all still supported with the Navigation API and now fire the "navigate" listener.
That is, they're also handled centrally.
Their signatures aren't modified in any way (i.e., they won't now return a `Promise`) by this new specification, and we imagine that in an older codebase, they'll be replaced by calls to `.navigate()` over time.

{% endAside %}

The `navigation.navigate()` method returns a object which contains two `Promise` instances in `{ committed, finished }`.
This allows the invoker can wait until either the transition is "committed" (the visible URL has changed and a new `NavigationEntry` is available) or "finished" (all promises passed to `transitionWhile()` are complete&mdash;or rejected, due to failure or being preempted by another navigation).

The `navigate` method also has an optional options object which controls how the navigation will occur.

> These options will allow you to `replace` the current URL, set a new immutable `state` (to be made available via the `.getState()` method on `NavigationEntry`s), and configure `navigateEvent.info`.

Reviewer 2: `replace` option got converted into `history: "replace"`, so this probably needs slight rephrasing.

It also includes `back()`, `forward()` and `reload()`.
These methods are all handled—just like `navigate()`—by the centralized "navigate" event listener.

### Form Submissions

Secondly, HTML `<form>` submission via POST is a special type of navigation, and the Navigation API can intercept it.
While it includes an additional payload, the navigation is still handled centrally by the "navigate" listener.

Form submission can be detected by looking for the `formData` property on the `NavigateEvent`.
Here's an example that simply turns any form submission into one which stays on the current page via `fetch()`:

```js
navigation.addEventListener('navigate', navigateEvent => {
  if (navigateEvent.formData && navigateEvent.canTransition) {
    // User submitted a POST form to a same-domain URL
    // (If canTransition is false, the event is just informative:
    // you can't intercept this request, although you could
    // likely still call .preventDefault() to stop it completely).

    const submitToServer = async () => {
      await fetch(navigateEvent.destination.url, {
        method: 'POST',
        body: navigateEvent.formData,
      });

>       // You could navigate again with {replace: true} to change the URL here,

Reviewer 2: replace: true -> history: "replace"

      // which might indicate "done"
    };
    navigateEvent.transitionWhile(submitToServer());
  }
});
```

## What's missing?

Despite the centralized nature of the "navigate" event listener, the current Navigation API specification doesn't trigger "navigate" on a page's first load.
And for sites which use [Server Side Rendering][ssr-definition] (SSR) for all states, this might be fine—your server could return the correct initial state, which is the fastest way to get content to your users.
But sites that leverage client-side code to create their pages may need to create an additional function to initialize their page.

> This is up for discussion [in the navigation-api repo][initial-event-discuss].

Reviewer 2: IMO we should remove this paragraph; I think we settled reasonably firmly on not including such an event.

- allow the user to complete their work (or go Back)
- remove a history entry on completion of a task

This could be perfect for temporary modals or interstitials: the new URL is something that a user can use the Back gesture to leave from, but they then cannot accidentally go Forward to open it again (because the entry has been removed).
This is just not possible with the current History API.

## Try the Navigation API

> You can try the Navigation API in Chrome 95 and above by enabling the "Experimental Web Platform features" flag.

Reviewer 2: Should we update this to Chrome 102 or above with no flags? Or figure out the version where all these renames landed (I think probably 101)?

You can also [try out a demo][demo] by [Domenic Denicola][domenic].

> We're especially eager for feedback on issues labelled with ["feedback wanted"][feedback-wanted] on GitHub.

Reviewer 2: [This list](https://github.com/WICG/navigation-api/issues?q=is%3Aissue+is%3Aopen+label%3A%22feedback+wanted%22) is pretty sparse these days, so maybe delete it.

Hero image from [Unsplash][hero-image], by [Jeremy Zero][hero-image-by].

[clunky-history-api]: https://html5doctor.com/interview-with-ian-hickson-html-editor/#:~:text=My%20biggest%20mistake%E2%80%A6there%20are%20so%20many%20to%20choose%20from!%20pushState()%20is%20my%20favourite%20mistake
[scroll-restoration]: /blog/history-api-scroll-restoration/
[image map]: https://developer.mozilla.org/docs/Web/HTML/Element/map
[back-forward-discuss]: https://github.com/WICG/navigation-api/issues/32
[loading-crbug]: https://bugs.chromium.org/p/chromium/issues/detail?id=1241202

> [abortable-fetch]: https://developers.google.com/web/updates/2017/09/abortable-fetch

Reviewer: ```suggestion
[abortable-fetch]: /blog/abortable-fetch/
```

> [ssr-definition]: https://developers.google.com/web/updates/2019/02/rendering-on-the-web#terminology

Reviewer: ```suggestion
[ssr-definition]: https://web.dev/rendering-on-the-web/#terminology
```
